schema.graphql

```diff
@@ -94,29 +98,68 @@ type UserExchangeData @entity {
     tokenFeesInUSD: BigDecimal!
 }
 
+enum EventType {
+    AddLiquidity,
+    RemoveLiquidity,
+    TokenPurchase,
+    EthPurchase
+}
+
 type Transaction @entity {
-    id: ID!                         # 0,1,2,3, etc....
-    tx: Bytes!
-    event: String!
+    id: ID!  # txn  hash concatenated with exchange - need this for token to token swaps
+    exchangeAddress: Bytes!
+    addLiquidityEvents: [AddLiquidityEvent!]
+    removeLiquidityEvents: [RemoveLiquidityEvent!]
+    tokenPurchaseEvents: [TokenPurchaseEvent!]
+    ethPurchaseEvents: [EthPurchaseEvent!]
     block: Int!
     timestamp: Int!
-    exchangeAddress: Bytes!
-    tokenAddress: Bytes!
-    tokenSymbol: String
     user: Bytes!
+    fee: BigDecimal!
+}
+
+# Note - No fee to provide liqidity
+type AddLiquidityEvent @entity {
```

Comment:
Yeah so the reason we added these is so that within a transaction you could see if there were multiple events - but i did add the interface. 

Happy to get on a call before we merge to decide on this one - i do agree it doesn't add a ton of value separating them into different entities as opposed to just one event entitiy

Reply:
Oh! I totally get it. That makes sense. I can show you another way to do it:

Update transaction like so:
```
type Transaction @entity {
    id: ID!  # txn  hash 
    exchangeAddress: Bytes!
    events: [Event!] @derivedFrom(field: "transaction")// where Event is querying on the interface so all 4 types of events should show up
    block: Int!
    timestamp: Int!
    user: Bytes!
    fee: BigDecimal!
}
```

Then after this, when you query a Tx it should show up multiple events. And you could also have a relation on the event, transaction derivedFrom ID. 

However this will change up how the front end queries the values. 

Also, it isn't immediately clear to me if this is helpful but you can split up events in the same transaction by taking the log index of the event. Then you can seperate the events by the index in the tx, which can be seen in the `EthereumEvent` type:
```
export class EthereumEvent {
  address: Address
  logIndex: BigInt
  transactionLogIndex: BigInt
  logType: string | null
  block: EthereumBlock
  transaction: EthereumTransaction
  parameters: Array<EthereumEventParam>
}
```
From `logIndex` or else within `EthereumTransaction`.

So what I recommend is we leave this fix out of the this pull request. And if you guys like it, I can make a PR after this is merged that implements it this way. Or else you guys can just keep it the way as it is working! Let me know

Reply:
Awesome. I say for now lets leave this out of the PR here. Then if you want to make that PR I'll take another look - I think you're rigth in that we can consolidate this a bit into less entities - this last solution would be good.